algorithms/sycl/Reduction.cpp

```diff
@@ -32,57 +27,98 @@ inline size_t getNearestPow2Number(size_t number) {
   return (BASE << power);
 }
 
-template <typename T, typename OperationT> void reduce(size_t size, size_t step, T *buffer, OperationT operation, void* streamPtr) {
 
-  auto targetSize = getNearestPow2Number(size / (step * 2));
-  auto rng = ::device::internals::computeDefaultExecutionRange1D(targetSize);
+template <typename T, typename OperationT>
+void reduce(T *to, T *from, size_t reducedSize, OperationT Operation, void* queuePtr) {
+  auto rng = ::device::internals::computeDefaultExecutionRange1D(reducedSize);
+  auto queue = reinterpret_cast<cl::sycl::queue*>(queuePtr);
+
+  queue->submit([&](handler &cgh) {
+    auto numLanes = ::device::internals::WARP_SIZE;
+    cl::sycl::local_accessor<T> shrMem{numLanes, cgh};
 
-  ((cl::sycl::queue *) streamPtr)->submit([&](handler &cgh) {
     cgh.parallel_for(rng, [=](nd_item<> item) {
-      int j = item.get_global_id(0) * (2 * step);
-      if (j < size) {
-        T buddyValue = operation.getDefaultValue();
-        T currentValue = buffer[j];
-
-        int buddyIndex = j + step;
-        if (buddyIndex < size) {
-          buddyValue = buffer[buddyIndex];
+      auto localRange = item.get_local_range();
+      auto gid = item.get_global_id(0);
+      auto lid = item.get_local_id(0);
+
+      shrMem[lid] = from[gid];
+      item.barrier();
+
+      for (int offset = numLanes/2; offset >= 1; offset /= 2) {
+        if (lid < offset) {
+          shrMem[lid] = Operation(shrMem[lid], shrMem[lid + offset]);
         }
-        buffer[j] = operation(currentValue, buddyValue);
+        item.barrier();
+      }
+
+      auto wid = item.get_group().get_id(0);
+      if (lid == 0) {
+        to[wid] = shrMem[0];
       }
     });
   });
 }
 
-template <typename T> T Algorithms::reduceVector(T *buffer, size_t size, const ReductionType type, void* streamPtr) {
-  if (api == nullptr)
+
+template <typename T>
+T Algorithms::reduceVector(T *buffer, size_t size, const ReductionType type, void* queuePtr) {
+  if (api == nullptr) {
     throw std::invalid_argument("api has not been attached to algorithms sub-system");
+  }
+
+  size_t adjustedSize = getNearestPow2Number(size);
+  const size_t totalBuffersSize = 2 * adjustedSize * sizeof(T);
+
+  T *reductionBuffer = reinterpret_cast<T *>(api->getStackMemory(totalBuffersSize));
```

Comment:
Hi. Yes, it would be safer, concise and more convenient.  However, we can not make it with templates because it would end up in headers which will eventually spread over seissol, and these headers will contain some types only known to `nvcc` `hipcc`, etc. compilers.

Currently all memory operations behaves like `malloc` or `memcpy` which operate on bytes.

In general, templates will work and simplify many things if SeisSol was a pure GPU project and we could stick to only one GPU vendor but we have different requirements: GPU part is optional